### PR TITLE
#345: Add error handling for product deletion

### LINKED ISSUE #345: Internal server error when deleting product that's in use
Create a product and add firmware from the CLI. Try to delete the product from the web. You'll get an internal server error.

**Expected behavior**
I was expecting an error, but something nicer that might have given me a hint that there was still something attached to the product.

### PR DESCRIPTION
**Pull request by chazwatkins**
* Add `Product.delete_changeset` with `no_assoc_constraint`
for `firmwares`
* Update `Products.delete_product` to use `Product.delete_changeset`
* Error flash message is shown when deleting a product with
associated firmwares for Web.
* Error response of `:conflict` with message is shown when
deleting a product with associated firmwares for the API.
* Add `NervesHubWWWWeb.FallbackController` to keep API and Web error handling the same.
* Update `NervesHubWWWWeb.FirmwareContoller.delete` and `NervesHubAPIWeb.FirmwareController.delete` to use `FallbackController`

closes #345

### CODE REVIEW

File: apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex [hunk 1]
```diff
@@ -38,6 +38,16 @@ defmodule NervesHubAPIWeb.ProductController do
   def delete(%{assigns: %{product: product}} = conn, _params) do
     with {:ok, %Product{}} <- Products.delete_product(product) do
       send_resp(conn, :no_content, "")
+    else
+      {:error, %Ecto.Changeset{errors: [firmwares: {reason, _}]}} ->
```
Comment (mobileoverlord) [inline apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex]: This seems like a pattern we could catch in the failback controller: https://github.com/nerves-hub/nerves_hub_web/blob/master/apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/fallback_controller.ex

File: apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex [hunk 2]
```diff
@@ -38,6 +38,16 @@ defmodule NervesHubAPIWeb.ProductController do
...
+        send_resp(conn, :conflict, reason)
+
+      {:error, _} ->
```
Comment (mobileoverlord) [inline apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex]: All other calls are also managed here: https://github.com/nerves-hub/nerves_hub_web/blob/master/apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/fallback_controller.ex#L31-L36

Comment (chazwatkins) [reply apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex]: I'll check into it and submit an update for firmware and products

Comment (chazwatkins) [reply apps/nerves_hub_api/lib/nerves_hub_api_web/controllers/product_controller.ex]: I've added the error logic to the FallbackController

**Comment by chazwatkins** [comment]
I've moved the error handling for `product` and `firmware` deletion to `FallbackController` as requested.  Let me know if there is anything else needed.

Thanks for taking the time to review the PR!

**Comment by mobileoverlord** [comment]
Merged via https://github.com/nerves-hub/nerves_hub_web/pull/599

**Pull request closed without merging**